### PR TITLE
Transfer Cont from @nomisrev to arrow

### PR DESCRIPTION
**Pull request by i-walker**
Thanks to @nomisRev proposal 🙌🏾
I've slightly changed the names from `Cont` to `Effect` and `ContEffect` to `EffectScope`, e.g.:

```kotlin
class DomainError(msg: String) 

context(CoroutineScope, EffectScope<DomainError>)
suspend fun program(): Int {
    val fa = async<Int> { shift(DomainError("error")) }
    val fb = async { delay(2000); 1 }
    return fa.await() + fb.await()
}
```

Feedback is welcome ☺️

### CODE REVIEW

**Review by raulraja — APPROVED** [review]
thanks @nomisRev @i-walker !

File: arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/Effect.kt
Comment (nomisRev) [inline arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/Effect.kt:104]:
```suggestion
  public fun attempt(): Effect<R, Result<A>> = effect {
```

All functions that return `Effect` should be non-suspending.

File: arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/Effect.kt
Comment (nomisRev) [inline arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/Effect.kt:112]:
```suggestion
  public fun handleError(recover: suspend (R) -> A): Effect<Nothing, A> = effect {
```

File: arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/Effect.kt
Comment (nomisRev) [inline arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/Effect.kt:116]:
```suggestion
  public fun <R2> handleErrorWith(recover: suspend (R) -> Effect<R2, A>): Effect<R2, A> = effect {
```

File: arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/Effect.kt
Comment (nomisRev) [inline arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/Effect.kt:120]:
```suggestion
  public fun <B> redeem(recover: suspend (R) -> B, transform: suspend (A) -> B): Effect<Nothing, B> =
```

File: arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/Effect.kt
Comment (nomisRev) [inline arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/Effect.kt:125]:
```suggestion
  public fun <R2, B> redeemWith(
```

File: arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/Effect.kt
Comment (nomisRev) [inline arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/Effect.kt:148]: Can we re-use the existing `Token` that is already in Arrow?

File: arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/Effect.kt
Comment (nomisRev) [inline arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/Effect.kt:208]:
```suggestion
public inline fun <R, A> effect(crossinline f: suspend EffectScope<R>.() -> A): Effect<R, A> =
```

File: arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/option.kt
Comment (nomisRev) [inline arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/option.kt:11]: I think we need`option { }` and `option.eager { }` to replace the existing behavior in `arrow.core.computations`.

File: arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/ior.kt
Comment (nomisRev) [inline arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/ior.kt:10]: We should also expose `object ior` with `suspend operator fun invoke` and `fun eager` like the other builders

File: EffectAndStructuredConcurrency.md
Comment (nomisRev) [inline EffectAndStructuredConcurrency.md:1]: I don't think this doc will show up on the Arrow Website this way.
Do we need to move it to the KDoc of the `Effect` interface? Or do we need to put it in a specific place and include it into the sidebar of the Arrow Core website?

**Review by nomisRev — CHANGES_REQUESTED** [review]
Thank you @i-walker for moving this code here 🙏
Couple of things still look incorrect. Great rename to `Effect` and `EffectScope` I like it!

File: arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/Effect.kt
Comment (i-walker) [reply arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/Effect.kt:148]: the only difference is that the one in arrow.generics is not transforming the hashCode with an unsigned Integer. If that's fine, I'll use it. Otherwise I could also change the one in arrow.generics?

Comment (nomisRev) [reply arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/Effect.kt:148]: Ah, it's in `arrow-continuations` 👍 
That module will be deprecated, and dissapear over time so lets keep this one instead.

It's not necessary to call `toUInt()` first and so we can improve this code with `override fun toString(): String = "Token(${hashCode().toString(16)})"`. It was needed in an older Kotlin version IIRC.

File: EffectAndStructuredConcurrency.md
Comment (i-walker) [reply EffectAndStructuredConcurrency.md]: It's now in Arrow Core and Fx

File: arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/option.kt
Comment (i-walker) [reply arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/option.kt]: I've exposed an eager version, but that would only point to eagerEffect with no dsl changes on top as in `option { }`
We could do it if the eagerEffect `shift` is not suspend.

File: arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/EagerEffectScope.kt
Comment (nomisRev) [inline arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/EagerEffectScope.kt:50]: This is redundant, since you can just use `value()` instead of `bind`.

File: arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/EffectScope.kt
Comment (nomisRev) [inline arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/EffectScope.kt:205]: idem

File: arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/option.kt
Comment (nomisRev) [inline arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/option.kt:55]: Why is this needed here?

File: arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/result.kt
Comment (nomisRev) [inline arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/result.kt:7]: We cannot expose this as a `public object`, see issue: https://github.com/arrow-kt/arrow/issues/2547

File: arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/result.kt
Comment (nomisRev) [inline arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/result.kt:20]: If we only have the `inline `version, we probably don't need the `object` indirection here.

File: arrow-libs/core/arrow-core/src/commonTest/kotlin/arrow/core/continuations/EagerEffectSpec.kt
Comment (nomisRev) [inline arrow-libs/core/arrow-core/src/commonTest/kotlin/arrow/core/continuations/EagerEffectSpec.kt:139]: Can we fix this by not using the method reference for `inline suspend` somewhere?

File: arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/option.kt
Comment (i-walker) [reply arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/option.kt:55]: removed it 👍🏾

File: arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/result.kt
Comment (i-walker) [reply arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/result.kt]: removed it 👍🏾

File: arrow-libs/core/arrow-core/src/commonTest/kotlin/arrow/core/continuations/EagerEffectSpec.kt
Comment (i-walker) [reply arrow-libs/core/arrow-core/src/commonTest/kotlin/arrow/core/continuations/EagerEffectSpec.kt]: yeah this was not failing

File: arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/option.kt
Comment (i-walker) [reply arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/option.kt]: Updated it with an dsl

File: arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/ior.kt
Comment (i-walker) [reply arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/ior.kt]: Done 👍🏾

File: arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/EagerEffect.kt
Comment (nomisRev) [inline arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/EagerEffect.kt:73]: I think we should use the same trick here as we do for `Effect`, so users have an escape hatch to differentiate between `JobCanncellationException` and `ShiftCanncelationException` when they have to in low-level niche use-cases.

File: arrow-site/docs/docs/arrow/core/continuations/README.md
Comment (nomisRev) [inline arrow-site/docs/docs/arrow/core/continuations/README.md:1]: This is not the documentation we want, this was part of the design process/documentation.
What we want instead is the [README.MD](https://github.com/nomisRev/Continuation/blob/main/README.md)

I'm thinking we should make it part of the KDoc on top of `Effect`, and link the page to `Effect` as well.

Currently, a lot of users have a hard time understanding the difference between `Effect` and `EagerEffect` and when or why they should be using `either { }` vs `either.eager { }`. This is currently completely undocumented inside `arrow-continuations`/`arrow.core.computations`.

I think we should add proper documentation in this PR before we merge, otherwise, it's going to be dropped to the bottom of the backlog again.

File: arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/EagerEffect.kt
Comment (i-walker) [reply arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/EagerEffect.kt:129]: added tests in EagerEffect and Effect for low-level use-cases

File: arrow-site/docs/docs/arrow/core/continuations/README.md
Comment (i-walker) [reply arrow-site/docs/docs/arrow/core/continuations/README.md:1]: Updated most places I was aware of, can I get another review

**Comment by i-walker** [comment]
I removed `EagerEffect<R, A>.toEffect(): Effect<R, A>`, because `effect { }` and `eagerEffect { }` can fold or `bind()` an `EagerEffect` value and it would introduce multiple paths of how to operate with this value doing something like `toEffect().bind()`, even though `bind()` is sufficient.

**Review by nomisRev — APPROVED** [review]
Great work @i-walker! Thanks for all the great effort moving this into Arrow 🥳 👏 👏 👏

File: arrow-site/docs/_data/sidebar-fx.yml
Comment (i-walker) [reply arrow-site/docs/_data/sidebar-fx.yml:39]: Makes sense 👍🏾

File: arrow-site/docs/_data/sidebar-core.yml
Comment (i-walker) [reply arrow-site/docs/_data/sidebar-core.yml:44]: Yes sounds great too, it would be great if we could link EffectScope with EagerEffectScope and showing they cover similar signatures. 
And rearranging it sounds super. 

On that note now that we moved on from Data type + extension (aka implementation of a Type class) we might want to consider just having Data types and computations. 

The few typeclasses we have left can also have a dedicated place, if necessary 
Just as a note, though.

File: arrow-site/docs/docs/fx/README.md
Comment (i-walker) [reply arrow-site/docs/docs/fx/README.md:66]: We have [this](https://arrow-kt.io/docs/fx/coroutines/), but it doesn't go as deep as `Semantics of Structured Concurrency and Effect` maybe we can create a ticket and I'd be happy to contribute it. Or help in whatever way

**Comment by i-walker** [comment]
Thanks for extensive review, learnt a lot 🤗

File: arrow-site/docs/docs/core/README.md
Comment (i-walker) [reply arrow-site/docs/docs/core/README.md:71]: yes we can leave it as a task in a ticket regarding how/ or what we should show on the site

**Pull request merged**
